### PR TITLE
puddletag: 2.0.1 -> 2.1.1

### PR DESCRIPTION
**Pull request by dschrempf**
Puddletag has changed the versioning procedure. Exact dependencies are in the requirements.txt but those are not provided by Nixpkgs. I jailbrake those exact dependency requirements (Puddletag runs fine here, but of course: Is there a better/preferrable way to work around this?).

###### Things done
- Add `chromaprint` to Python packages (new dependency).
- Update `puddletag`.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dschrempf** [comment]
@peterhoeg

**Comment by dschrempf** [comment]
Sorry, I just saw that the formatter changed some other stuff. I am going to revert those formatting changes now.

**Comment by dschrempf** [comment]
Now, that's better.

**Comment by peterhoeg** [comment]
I have a patch locally that I haven't pushed with a different approach.

Instead of using the locked dependencies, we copy out requirements.in (which only has a few hard requirements) to requirements.txt and otherwise pass it the versions we have in nixpkgs. It's been working fine here.

**Comment by peterhoeg** [comment]
And I just realized that you cannot attach a patch file to when replying via email so here goes:

https://gist.github.com/7533042e681287dbebb1434bdaf5644e

File: pkgs/development/python-modules/chromaprint/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, fetchFromGitHub
+, python3
+}:
+
+let
+  pname = "chromaprint";
```
Comment (peterhoeg) [inline pkgs/development/python-modules/chromaprint/default.nix:7]: you can move pname and version into the buildPythonPackage call as it's being called with a recursive set anyway.

File: pkgs/development/python-modules/chromaprint/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, fetchFromGitHub
+, python3
```
Comment (peterhoeg) [inline pkgs/development/python-modules/chromaprint/default.nix:3]: python modules are not given a specific python version as they can be built against multiple versions. Instead guard on versions.

**Review by peterhoeg — CHANGES_REQUESTED** [review]

**Comment by dschrempf** [comment]
@peterhoeg, thank you for your review! If you have a working version that you prefer, why not use yours? I cannot apply the patch here, because my file has diverged, and I don't really want to copy over the stuff by hand. Let me know how you want to proceed!

More generally, it would be great to have better ways of avoiding duplicate work in Nixpkgs...

**Comment by peterhoeg** [comment]
> If you have a working version that you prefer, why not use yours?
> More generally, it would be great to have better ways of avoiding duplicate work in Nixpkgs...

I was lazy.... I should obviously have pushed it upstream, but put it off as I wanted to write some comments and needed to split out chromaprint.

Not great reasons I admit, but thank you for the push to do it.

One thing, just because I happen to be maintaining puddletag in nixpkgs doesn't mean that my arguments carry more weight, so I would like to hear your thoughts on my approach. I'm not married to it, but I think it makes more sense because:

1. we're separating the "override versions because upstream wants it" and "override versions because it happens to be what upstream was built against"

2. I think it's clearer when defining a variable that contains what we are overriding, rather than defining it in the phase where we are acting on it. In my approach postPatch doesn't contain any data - only the actions on said data.

It's not that one is inherently better - I just personally subscribe to the "principle of least surprise".

Lastly, I took the liberty of adding you in as a maintainer for puddletag. Hope you're OK with that.

**Comment by dschrempf** [comment]
I agree with your points. I was not very satisfied with my solution, that's also why I asked in the beginning.

My comment about the Nixpkgs workflow was more of general nature, I hope you didn't take any offense!

**Comment by peterhoeg** [comment]
> My comment about the Nixpkgs workflow was more of general nature, I hope you didn't take any offense!

Oh, none at all.

We're stretched pretty thin in terms of resources so having people duplicate efforts just because I couldn't be bothered to the right thing is of course annoying for everyone else. Mea culpa.

**Pull request closed without merging**